### PR TITLE
FLE implementation with property value converter.

### LINKED ISSUE #763: Field Level Encryption [DATACOUCH-455]
**[Domenico Ursoleo](https://jira.spring.io/secure/ViewProfile.jspa?name=udomenico)** opened **[DATACOUCH-455](https://jira.spring.io/browse/DATACOUCH-455?redirect=false)** and commented

Couchbase has a Field Level Encryption feature that encrypt field of a Document simply putting the `@EncryptField` annotation on a field of the document model and addind a CryptoManager to the environment,

With this configuration:

```java
@Configuration
@EnableCouchbaseRepositories(basePackages = { "com.example.demo.db" })
public class MyCouchbaseConfig extends AbstractCouchbaseConfiguration {
 
@Override 
protected CouchbaseEnvironment getEnvironment() { 
try {
 JceksKeyStoreProvider kp = new JceksKeyStoreProvider("secret");
 kp.publicKeyName("mypublickey");
 kp.storeKey("mypublickey",
           "!mysecretkey#9^5usdk39d&dlf)03sL".getBytes(Charset.forName("UTF-8")));
 kp.signingKeyName("HMACsecret");
 kp.storeKey("HMACsecret", "myauthpassword".getBytes(Charset.forName("UTF-8")));
 AES256CryptoProvider aes256CryptoProvider = new AES256CryptoProvider(kp);
 CryptoManager cryptoManager = new CryptoManager();
 cryptoManager.registerProvider("MyAESProvider", aes256CryptoProvider);
 return 
    DefaultCouchbaseEnvironment.builder().cryptoManager(cryptoManager).build();
 } catch (Exception ex) { return null; } }
}
```

 Given the model

 

```java
@Document
public class User {
@Id
@GeneratedValue(strategy = GenerationStrategy.UNIQUE)
private String id;

@Field("user")
private String userName;

@EncryptedField(provider = "MyAESProvider") 
@Field("pwd") 
private String password;

// Setters & getters
}
```

Persisting it via a simple CrudRepository should encrypt the password like calling the native couchbase java client:

```java
EntityDocument<User> document = EntityDocument.create(user);          bucket.repository().upsert(document);
 
```

 

 

---

**Affects:** 3.1.8 (Lovelace SR8)

1 votes, 2 watchers

### PR DESCRIPTION
**Pull request by mikereiche**
Closes #763

- [x] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [x] There is a ticket in the bug tracker for the project in our [JIRA](https://jira.spring.io/browse/DATACOUCH).
- [x] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

### CODE REVIEW

File: spring-data-couchbase/src/main/java/org/springframework/data/couchbase/core/convert/AbstractCouchbaseConverter.java [hunk 1]
```diff
@@ -93,39 +98,70 @@ public void afterPropertiesSet() {
...
+		Object result = this.conversions.getCustomWriteTarget(prop.getType()) //
+				.map(it -> this.conversionService.convert(value, new TypeDescriptor(prop.getField()),
+						TypeDescriptor.valueOf(it))) //
 				.orElseGet(() -> Enum.class.isAssignableFrom(value.getClass()) ? ((Enum<?>) value).name() : value);
 
+		return result;
+
 	}
 
-	/* TODO needed later
```
Comment (mikereiche) [inline spring-data-couchbase/src/main/java/org/springframework/data/couchbase/core/convert/AbstractCouchbaseConverter.java:108]: The 'diff' gets confusing here.  I removed a commented-out convertToCouchbaseType() method here and added a new convertForWriteIfNeeded() (above) that has three args vs. the existing one that had only one arg (below).

File: spring-data-couchbase/src/main/java/org/springframework/data/couchbase/core/convert/CryptoConverter.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+		}
+		// it's decrypted into a String. Now figure out how to convert to the property type.
+		CouchbaseConversionContext ctx = (CouchbaseConversionContext) context;
+		CouchbasePersistentProperty property = ctx.getProperty();
+		org.springframework.data.convert.CustomConversions conversions = ctx.getConverter().getConversions();
+		ConversionService svc = ctx.getConverter().conversionService;
+
+		boolean wasString = false;
+		List<Exception> exceptionList = new LinkedList<>();
+		String decryptedString = new String(decrypted);
```
Comment (mikereiche) [inline spring-data-couchbase/src/main/java/org/springframework/data/couchbase/core/convert/CryptoConverter.java:65]: I'm not really happy with the following code that takes the decrypted string and turns it into a simple type.  But it seems to do the job.

File: spring-data-couchbase/src/main/java/org/springframework/data/couchbase/core/convert/CryptoConverter.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+		byte[] plainText;
+		CouchbaseConversionContext ctx = (CouchbaseConversionContext) context;
+		CouchbasePersistentProperty property = ctx.getProperty();
+		org.springframework.data.convert.CustomConversions conversions = ctx.getConverter().getConversions();
+
+		Class<?> sourceType = context.getProperty().getType();
+		Class<?> targetType = conversions.getCustomWriteTarget(context.getProperty().getType()).orElse(null);
+
+		value = ctx.getConverter().getPotentiallyConvertedSimpleWrite(property, ctx.getAccessor(), false);
+		if (conversions.isSimpleType(sourceType)) {
```
Comment (mikereiche) [inline spring-data-couchbase/src/main/java/org/springframework/data/couchbase/core/convert/CryptoConverter.java:144]: Similar comment to above - I'm not really happy with this code that converts the value into a string for decryptiing.  But it seems to work.

File: spring-data-couchbase/src/main/java/org/springframework/data/couchbase/core/convert/DateConverters.java [hunk 4]
```diff
@@ -204,7 +205,7 @@ public enum NumberToDateTimeConverter implements Converter<Number, DateTime> {
 
 		@Override
 		public DateTime convert(Number source) {
-			return source == null ? null : new DateTime(source.longValue());
+			return source == null ? null : new DateTime(source.longValue(), DateTimeZone.UTC);
```
Comment (mikereiche) [inline spring-data-couchbase/src/main/java/org/springframework/data/couchbase/core/convert/DateConverters.java:208]: This change is to make the Couchbase deserialization compatible with the SDK (via jackson-jodatime) which is in UTC (vs. local time zone).

File: spring-data-couchbase/src/test/java/org/springframework/data/couchbase/repository/CouchbaseRepositoryFieldLevelEncryptionIntegrationTests.java [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+			if (config().isUsingCloud()) {
+				builder.securityConfig(
+						SecurityConfig.builder().trustManagerFactory(InsecureTrustManagerFactory.INSTANCE).enableTls(true));
+			}
+			CryptoManager cryptoManager = cryptoManager();
+			builder.cryptoManager(cryptoManager).build();
+		}
+
+		@Override
+		protected CryptoManager cryptoManager() {
```
Comment (mikereiche) [inline spring-data-couchbase/src/test/java/org/springframework/data/couchbase/repository/CouchbaseRepositoryFieldLevelEncryptionIntegrationTests.java:183]: This is the method the user will need to override.

File: spring-data-couchbase/src/test/java/org/springframework/data/couchbase/domain/UserEncrypted.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+	public UserEncrypted(final String id, final String firstname, final String lastname, final String encryptedField) {
+		this();
+		this.id = id;
+		this.firstname = firstname;
+		this.lastname = lastname;
+		this.encryptedField = encryptedField;
+	}
+
+	@Version protected long version;
+	@Encrypted public String encryptedField;
```
Comment (mikereiche) [inline spring-data-couchbase/src/test/java/org/springframework/data/couchbase/domain/UserEncrypted.java:77]: Normally spring-data would use an @ValueConverter annotation for this. But my CouchbaseValueConverterFactory also recognizes Couchbase @Encrypted

**Review by dnault — APPROVED** [review]
+2, after the improvements we chatted about today

**Pull request merged**